pkg/ipcache/integration_test.go

```diff
@@ -744,3 +745,156 @@ func TestServerProxyClientIntegration(t *testing.T) {
 		return !found
 	}, 15*time.Second)
 }
+
+// Helper to calculate the size of a directory
+func dirSize(path string) (int64, error) {
+	var size int64
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			size += info.Size()
+		}
+		return err
+	})
+	return size, err
+}
+
+// TestScalability measures propagation delay, memory, and disk usage.
+func TestScalability(t *testing.T) {
+	const (
+		numClients          = 1
+		numNamespaces       = 1
+		numPodsPerNamespace = 1000000
+		totalPods           = numNamespaces * numPodsPerNamespace
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// 1. Setup server
+	server, _, cleanup := setupTest(t, ctx)
+	defer cleanup()
+	listenURL := server.listenURL
+
+	// 2. Setup clients
+	clients := make([]*Client, numClients)
+	for i := 0; i < numClients; i++ {
+		dbDir := filepath.Join(t.TempDir(), fmt.Sprintf("ipcache%d.bolt", i))
+		boltStore, err := NewBoltStore(dbDir)
+		if err != nil {
+			t.Fatalf("Failed to create bolt store: %v", err)
+		}
+		lruStore := NewLRUStore(boltStore, totalPods*2)
+		client, err := NewClient(context.Background(), listenURL, nil, lruStore, boltStore, fmt.Sprintf("node-%d", i))
+		if err != nil {
+			t.Fatalf("Failed to create client %d: %v", i, err)
+		}
+		defer client.Close()
+		clients[i] = client
+	}
+
+	// 3. Create namespaces and pods
+	ips := make([]string, totalPods)
+	podInfos := make([]*api.PodInfo, totalPods)
+	for i := 0; i < numNamespaces; i++ {
+		ns := fmt.Sprintf("ns%d", i)
+		for j := 0; j < numPodsPerNamespace; j++ {
+			podIndex := i*numPodsPerNamespace + j
+			podName := fmt.Sprintf("pod-%d", podIndex)
+			ip := make(net.IP, 4)
+			binary.BigEndian.PutUint32(ip, uint32(podIndex))
+			ips[podIndex] = ip.String()
+			podInfos[podIndex] = &api.PodInfo{Name: podName,
+				Labels: map[string]string{"foo": "bar"},
+				Namespace: &api.Namespace{
+					Name:   ns,
+					Labels: map[string]string{"foo": "bar"},
+				},
+			}
+		}
+	}
+
+	// 4. Measure initial state
+	var beforeMemStats runtime.MemStats
+	runtime.ReadMemStats(&beforeMemStats)
+	serverDirSize, err := dirSize(server.etcd.Config().Dir)
+	if err != nil {
+		t.Fatalf("Failed to get server dir size: %v", err)
+	}
+	t.Logf("Initial server disk usage: %.2f KB", float64(serverDirSize)/1024)
+
+	// 5. Upsert all pods and measure propagation delay
+	var totalPropagationDelay time.Duration
+	var propagationDelayCount int64
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+
+	for i := 0; i < totalPods; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			startTime := time.Now()
+			if err := server.Upsert(ips[i], podInfos[i]); err != nil {
```

Comment:
Creating all pods at once is not very realistic and causes a huge burst that will overload etcd. Etcd without K8s api doesn't have any overload protection like APF. Better to do a setup all pods and then keep updating them with some rate like >100 QPS.